### DISCUSSION TITLE
How do I ensure alignment between Go and C structs?

### DISCUSSION THREAD

**Discussion by binw666** [cilium/ebpf#857] [Q&A] [answered]
How are the items in struct of go associated with the items in struct of c? By the order of the items in struct or by the name of the items in the struct. When I try to use `Update` to update the value in the map by key, I get can't `marshal value: main.ruleIpTcp doesn't marshal to 28 bytes`. I haven't found any examples about `Update`. I don't know what I'm doing wrong.
The definition of map in c : 
```c
struct ip_tcp_rule{
    __u8 flag;
    __u32 src_ip;
    __u8 src_ip_mask;
    __u32 dst_ip;
    __u8 dst_ip_mask;
    __u8 min_ttl;
    __u8 max_ttl;
    __u16 src_port_min;
    __u16 src_port_max;
    __u16 dst_port_min;
    __u16 dst_port_max;
};

struct bpf_map_def SEC("maps") ip_tcp_rules_0 =
{
    .type = BPF_MAP_TYPE_ARRAY,
    .key_size = sizeof(__u32),
    .value_size = sizeof(struct ip_tcp_rule),
    .max_entries = MAX_RULE_IP_TCP_NUM
};
```
code in go
```go
type ruleIpTcp struct {
	Flag       uint8  `json:"flag"`
	SrcIp      uint32 `json:"src_ip"`
	SrcIpMask  uint8  `json:"src_ip_mask"`
	DstIp      uint32 `json:"dst_ip"`
	DstIpMask  uint8  `json:"dst_ip_mask"`
	MinTTL     uint8  `json:"min_ttl"`
	MaxTTL     uint8  `json:"max_ttl"`
	SrcPortMin uint16 `json:"src_port_min"`
	SrcPortMax uint16 `json:"src_port_max"`
	DstPortMin uint16 `json:"dst_port_min"`
	DstPortMax uint16 `json:"dst_port_max"`
}

// ...
var v ruleIpTcp
ui := uint32(1)
// ... 
err = opt_map.Update(ui, v, ebpf.UpdateAny)
```

**Comment by binw666** [comment]
I find the problem. I think that the size of the struct in go is 21, but the size of the struct in c is 28. So the items in struct of go associated with the items in struct of c by the order of items. When the struct is padded both in c and go, it works well.

**Comment by ti-mo** [comment, marked as answer]
There are many resources and tools available to work with Go struct alignment, and yes, it's something you need to pay attention to when sharing them between C and Go. The Go struct you pasted here seems to be 28 bytes long on a 64-bit architecture: https://go.dev/play/p/u4X0Vn-finI. Hint: run it through [`structslop`](https://medium.com/orijtech-developers/efficient-struct-packing-guided-pass-for-go-92255872ec72) to optimize its size.

The following C program also evaluates to 28:
```c
#include <stdio.h>

typedef unsigned char __u8;
typedef short unsigned int __u16;
typedef unsigned int __u32;
typedef long long unsigned int __u64;

struct ip_tcp_rule{
    __u8 flag;
    __u32 src_ip;
    __u8 src_ip_mask;
    __u32 dst_ip;
    __u8 dst_ip_mask;
    __u8 min_ttl;
    __u8 max_ttl;
    __u16 src_port_min;
    __u16 src_port_max;
    __u16 dst_port_min;
    __u16 dst_port_max;
};

int main() {
   printf("%ld", sizeof(struct ip_tcp_rule));
   return 0;
}
```

```
λ  ~  clang main.c && ./a.out
28
```

Clang's bpf backend may align fields differently for some reason, I'm not exactly sure why.

You can use `CollectionSpec.Types` to inspect the BTF info of your C structs and make sure the fields align, see cilium/cilium for inspiration: https://github.com/cilium/cilium/pull/20809. This functionality might come to the library in the distant future, not sure when yet.
